python/shotgun_model/data_handler_nav.py

```diff
@@ -0,0 +1,230 @@
+            of field names to return.
+
+        :param str cache_path: Path to cache file location.
+
+        :param :class:`~PySide.QtGui.QObject` parent: Parent QT object.
+        """
+        super(ShotgunNavDataHandler, self).__init__(cache_path, parent)
+        self.__root_path = root_path
+        self.__seed_entity_field = seed_entity_field
+        self.__entity_fields = entity_fields
+
+    def generate_data_request(self, data_retriever, path):
+        """
+        Generate a data request for a data retriever.
+
+        Once the data has arrived, the caller is expected to
+        call meth:`update_data` and pass in the received
+        data payload for processing.
+
+        :param data_retriever: :class:`~tk-framework-shotgunutils:shotgun_data.ShotgunDataRetriever` instance.
+        :returns: Request id or None if no work is needed
+        """
+        self._log_debug("generate_data_request for path %s" % path)
+
+        worker_id = data_retriever.execute_nav_expand(
+            path,
+            self.__seed_entity_field,
+            self.__entity_fields
+        )
+
+        return worker_id
+
+    @log_timing
+    def update_data(self, sg_data):
+        """
+        The counterpart to :meth:`generate_data_request`. When the data
+        request has been carried out, this method should be called by the calling
+        class and the data payload from Shotgun should be provided via the
+        sg_data parameter.
+
+        The shotgun nav data is compared against an existing part of the tree and
+        a list of differences is returned, indicating which nodes were
+        added, deleted and modified, on the following form::
+
+            [
+             {
+                "data": ShotgunDataItem instance,
+                "mode": self.UPDATED|ADDED|DELETED
+             },
+             {
+                "data": ShotgunDataItem instance,
+                "mode": self.UPDATED|ADDED|DELETED
+             },
+             ...
+            ]
+
+        :param sg_data: list, resulting from a Shotgun nav_expand query
+        :returns: list of updates. see above
+        :raises: :class:`ShotgunModelDataError` if no cache is loaded into memory
+        """
+        if self._cache is None:
+            raise ShotgunModelDataError("No data currently loaded in memory!")
+
+        self._log_debug("Updating %s with %s shotgun records." % (self, len(sg_data)))
+
+        item_path = sg_data.get(self._SG_PATH_FIELD, None)
+
+        self._log_debug("Got hierarchy data for path: %s" % (item_path,))
+
+        if not item_path:
+            raise ShotgunModelDataError(
+                "Unexpected error occurred. Could not determine the path"
+                "from the queried hierarchy item."
+            )
+
+        if len(self._cache[self.CACHE_CHILDREN]) == 0:
+            self._log_debug("In-memory cache is empty.")
+
+        # ensure the data is clean
+        self._log_debug("sanitizing data...")
```

Comment:
i thought about it - it's actually not that much commonality - mostly init stuff, so i left it separate. but i wasn't sure.